common/types/malfeasance.go

```diff
@@ -0,0 +1,201 @@
+package types
+
+import (
+	"github.com/spacemeshos/go-scale"
+
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/log"
+)
+
+//go:generate scalegen -types MalfeasanceGossip,MultiATXsMsg,MultiBallotsMsg,HareEquivocationMsg,HareMetadata
+
+type MalfeasanceType uint16
+
+const (
+	MultipleATXs MalfeasanceType = iota + 1
+	MultipleBallots
+	HareEquivocation
+)
+
+type MalfeasanceProof struct {
+	// for network upgrade
+	Layer LayerID
+	Type  MalfeasanceType
+	// conflicting messages signed by the same NodeID
+	Messages []any
+}
+
+func (t *MalfeasanceProof) EncodeScale(enc *scale.Encoder) (int, error) {
```

Comment:
@dshulyak please advise if improvements can be made in the custom encode/decode. thank you

Reply:
i would use a following approach:

1. define different malfeasence proofs. all of them can be autogenerated

```go
type AtxProof struct {
    [2]ConcreteAtxMessage
}

type HareProof struct {
   [2]ConcreteHareMessage
}
```

2. define enum for them. this is the only object that needs custom encoding, as the library doesn't support autogenerating it and it is not straighforward, but in general should be possible to autogenerate it with some work

```go

type ProofEnum struct {
    Type uint8
    Proof scale.Type
}

func (e *ProofEnum) EncodeScale(e *scale.Encode) {
    // not compact, as scale spec uses "full" uint8 for enums
    scale.EncodeByte(e, byte(e.Type))
   e.Proof.EncodeScale(e)
}

func (e *ProofEnum) DecodeScale(d *scale.Decoder) {
   typ := scale.DecodeByte(d)
   switch typ {
      // handle concrete type
   }
}
```  

3. higher level malfeasence proof for versioning. autogenerated

```go
type MalfeasenceProof struct {
    Layer types.LayerID
    Enum ProofEnum
}

```

Reply:
@dshulyak done as suggested with some name changes. thanks for helping.